### PR TITLE
swap holder

### PR DESCRIPTION
**Pull request by bekauz**
base implementation for #103 

review after #102

### CODE REVIEW

File: contracts/swap-holder/src/msg.rs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+#[cw_serde]
+pub struct InstantiateMsg {
+    /// Address for the clock. This contract verifies
+    /// that only the clock can execute Ticks
+    pub clock_address: String,
+    /// address of the next contract to forward the funds to.
+    /// usually expected tobe the splitter.
+    pub next_contract: String,
+    /// block height of covenant expiration. Position is exited
```
Comment (uditvira) [inline contracts/swap-holder/src/msg.rs:14]: To clarify, we can also accept lockup config as a timestamp

File: contracts/swap-holder/src/msg.rs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    /// automatically upon reaching that height.
+    pub lockup_config: LockupConfig,
+    /// parties engaged in the POL.
```
Comment (uditvira) [inline contracts/swap-holder/src/msg.rs:17]:
```suggestion
    /// parties engaged in the swap.
```

**Review by uditvira — COMMENTED** [review]
I'm unclear what the purpose of the splitter is in the 2 party swap scenario. The holder takes care of holding and refunding the funds. Splitting the funds should be a trivial function for it, correct?

Can we justify this design choice?

File: packages/covenant-utils/src/lib.rs [hunk 3]
```diff
@@ -116,4 +134,209 @@ pub mod neutron_ica {
...
+        }
+    }
+}
+
+#[cw_serde]
+pub struct CovenantParty {
+    /// authorized address of the party
+    pub addr: Addr,
+    /// denom provided by the party
+    pub provided_denom: String,
```
[the remark this answers is not in the document]
Comment (bekauz) [reply packages/covenant-utils/src/lib.rs:260]: For now there is no specific reason. Just felt like at least in this covenant, a party will always be associated with/identified by its respective denom.

Worth to mention that I am somewhat intentionally leaving some ambiguity in the interface design at this stage. Once more covenants are finished, I expect some patterns to appear and we can standardise everything.

**Comment by bekauz** [comment]
> I'm unclear what the purpose of the splitter is in the 2 party swap scenario. The holder takes care of holding and refunding the funds. Splitting the funds should be a trivial function for it, correct?

Great point. It is indeed trivial to directly send the funds to routers with no splitter. I opted to use the splitter here because I see it as our dedicated module for distributing the funds, and there's probably some value in reusing components as much as we can.

That being said, do you think it's worth to get rid of it? We would save a trivial amount of gas and have one less contract involved. 

cc @Art3miX

**Pull request closed without merging**